src/chan_vese.jl

```diff
@@ -0,0 +1,258 @@
+```
+
+# References
+
+[1] An Active Contour Model without Edges, Tony Chan and Luminita Vese, 
+    Scale-Space Theories in Computer Vision, 1999, :DOI:`10.1007/3-540-48236-9_13`
+[2] Chan-Vese Segmentation, Pascal Getreuer Image Processing On Line, 2 (2012), 
+    pp. 214-224, :DOI:`10.5201/ipol.2012.g-cv`
+[3] The Chan-Vese Algorithm - Project Report, Rami Cohen, 2011 :arXiv:`1107.2782`
+"""
+function chan_vese(img::GenericGrayImage;
+                   μ::Float64=0.25,
+                   λ₁::Float64=1.0,
+                   λ₂::Float64=1.0,
+                   tol::Float64=1e-3,
+                   max_iter::Int64=500,
+                   Δt::Float64=0.5,
+                   reinitial_flag::Bool=false)
+    # Signs used in the codes and comments mainly follow paper[3] in the References.
+    img = float64.(channelview(img))
+    iter = 0
+    h = 1.0
+    del = tol + 1
+    img .= img .- minimum(img)
+
+    if maximum(img) != 0
+        img .= img ./ maximum(img)
+    end
+
+    # Precalculation of some constants which helps simplify some integration   
+    area = length(img) # area = ∫H𝚽 + ∫H𝚽ⁱ
+    ∫u₀ = sum(img)     # ∫u₀ = ∫u₀H𝚽 + ∫u₀H𝚽ⁱ
+
+    # Initialize the level set
+    𝚽ⁿ = initial_level_set(size(img))
+
+    # Preallocation and initializtion
+    H𝚽 = trues(size(img)...)
+    𝚽ⁿ⁺¹ = similar(𝚽ⁿ)
+
+    # The upper bounds of 𝚽ⁿ's coordinates is `m` and `n`
+    s, t = first(CartesianIndices(𝚽ⁿ))[1], first(CartesianIndices(𝚽ⁿ))[2]
+    m, n = last(CartesianIndices(𝚽ⁿ))[1], last(CartesianIndices(𝚽ⁿ))[2]
+    
+    while (del > tol) & (iter < max_iter)
+        ϵ = 1e-8
+        diff = 0
+
+        # Calculate the average intensities
+        @. H𝚽 = 𝚽ⁿ > 0 # Heaviside function
+        c₁, c₂ = calculate_averages(img, H𝚽, area, ∫u₀) # Compute c₁(𝚽ⁿ), c₂(𝚽ⁿ)
+
+        # Calculate the variation of level set 𝚽ⁿ
+        for idx in CartesianIndices(𝚽ⁿ) # Denote idx = (x, y)
+            # i₊ ≔ i₊(x, y), denotes 𝚽ⁿ(x, y + 1)'s CartesianIndex
+            # j₊ ≔ j₊(x, y), denotes 𝚽ⁿ(x + 1, y)'s CartesianIndex
+            # i₋ ≔ i₋(x, y), denotes 𝚽ⁿ(x, y - 1)'s CartesianIndex
+            # j₋ ≔ j₋(x, y), denotes 𝚽ⁿ(x - 1, y)'s CartesianIndex
+            # Taking notice that if 𝚽ⁿ(x, y) is the boundary of 𝚽ⁿ, than 𝚽ⁿ(x ± 1, y), 𝚽ⁿ(x, y ± 1) might be out of bound.
+            # So the pixel values of these outbounded terms are equal to 𝚽ⁿ(x, y)
+            i₊ = idx[2] != n ? idx + CartesianIndex(0, 1) : idx
+            j₊ = idx[1] != m ? idx + CartesianIndex(1, 0) : idx
+            i₋ = idx[2] != t ? idx - CartesianIndex(0, 1) : idx
+            j₋ = idx[1] != s ? idx - CartesianIndex(1, 0) : idx
+
+            𝚽₀  = 𝚽ⁿ[idx] # 𝚽ⁿ(x, y)
+            u₀ = img[idx] # u₀(x, y)
+            𝚽ᵢ₊ = 𝚽ⁿ[i₊] # 𝚽ⁿ(x, y + 1)
+            𝚽ⱼ₊ = 𝚽ⁿ[j₊] # 𝚽ⁿ(x + 1, y)
+            𝚽ᵢ₋ = 𝚽ⁿ[i₋] # 𝚽ⁿ(x, y - 1)
+            𝚽ⱼ₋ = 𝚽ⁿ[j₋] # 𝚽ⁿ(x - 1, y)
```

Comment:
The current version hardcodes to 2-dimensional and it seems unnecessary. How about:

```julia
julia> function direct_neighborhoods(p::CartesianIndex{N}) where N
           neighbors = ntuple(N) do i
               Δ = CartesianIndex(ntuple(j->Int(i==j), N))
               p-Δ, p+Δ
           end
           reduce(neighbors) do i, j
               (i..., j...)
           end
       end
direct_neighborhoods (generic function with 1 method)

julia> direct_neighborhoods(CartesianIndex(2, 3))
(CartesianIndex(1, 3), CartesianIndex(3, 3), CartesianIndex(2, 2), CartesianIndex(2, 4))

julia> max.(Ref(CartesianIndex(1, 1)), direct_neighborhoods(CartesianIndex(1, 2)))
(CartesianIndex(1, 2), CartesianIndex(2, 2), CartesianIndex(1, 1), CartesianIndex(1, 3))
```

and I believe it's possible to solve equation (9) for generic n-dimensional case.

Reply:
@JKay0327, I know it takes a little while to learn how to write dimension-agnostic code, so feel free to ask questions. But it's worth it for two reasons: (1) it's one of the coolest features of Julia, and (2) future-you will be grateful when you have that MRI taken of your knee and your orthopedic surgeon uses the Chan-Vese algorithm you wrote to plan the surgical approach :smile:.